File: linked_list_values/linked_list_values.py

# recursive
#O(n) time | O(n) space where n is the number of nodes in the linked_list
def linked_list_values(head):
  values = []
  fill_values(head, values)
  return values
  
def fill_values(head, values):
  if head is None: 
    return 
  values.append(head.val) 
  fill_values(head.next, values)
  
# iterative
# def linked_list_values(head):
#   list = []
#   current = head
#   while current is not None:
#     list.append(current.val)
#     current = current.next
    
#   return list
​
​
# print(linked_list_values(a))

File: linked_list_values/linked_list_values.js
const a = new Node("a");
const b = new Node("b");
const c = new Node("c");
const d = new Node("d");


a.next = b;
b.next = c;
c.next = d;


// a -> b -> c -> d


linkedListValues(a); // -> [ 'a', 'b', 'c', 'd' ]
const x = new Node("x");
const y = new Node("y");


x.next = y;


// x -> y


linkedListValues(x); // -> [ 'x', 'y' ]
const q = new Node("q");


// q


linkedListValues(q); // -> [ 'q' ]
linkedListValues(null); // -> [ ]
// class Node {
//   constructor(val) {
//     this.val = val;
//     this.next = null;
//   }
// }


//O(n) time and O(n) space
const linkedListValues = (head) => {
  const listVals = [];
  linkedListValuesHelper(head, listVals);
  return listVals;
};


const linkedListValuesHelper = (head, listVals) => {
  if (!head) return [];
  listVals.push(head.val);
  linkedListValuesHelper(head.next, listVals);
}


module.exports = {
  linkedListValues,
};

